### ISSUE TITLE
[Enhancement][move compiler] Emit MoveLoc more aggressively

### ISSUE THREAD

**Issue by sblackshear** [diem/diem#8257]
## Explanation/Repro
```
// Test.move
address 0x1 {
module Test {
  public fun id(addr: address): address {
      addr
  }
}
}
```
Repro:
```
move publish Test.move
move view storage/0x00000000000000000000000000000001/modules/Test.mv
module 1.Test {
public id(): address {
	0: CopyLoc[0](Arg0: address)
	1: Ret
}
}
```
We'd like the compiler to emit a `MoveLoc` instead of a `CopyLoc` here, and it does for other very similar programs (e.g., switch `address` to `vector<u8>`, `fun id<T: copy>(t: T): T`).

## Explanation of current behavior

`copy` is used for any primitive/base type (not `vector` or `signer`) without annotation. This includes references. move is used for any other position without annotation
References are the only value that gets moved for its last usage. This is what you rightfully expected in the example, but it turns out it is really difficult to do for anything besides references because the local could be borrowed.
Consider
```
let x = 0;
let r = &x;
let y = x;
*r
```
The last usage of `x` is in `let y = x`, so a simple (non-borrow aware) liveness analysis would try to make `let y = move x` since it is the last usage.
But for references, making this a move is safe without being aware of borrows!
```
let x: &S = ...;
let r = &x.f;
let y = /* move */ x; // Valid move!
*r
```
And in some cases, this move for references is necessary for the system to feel “correct”. For example
```
let x: &mut S = ...;
let r = &mut x.f;
*r = *r + 1;
foo(x)
```
If `*r = *r + 1` was not `*(move r) = *(copy r) + 1` this would lead to some confusing errors about `x` being borrowed.
TLDR: We originally wanted to make the last usage always a `MoveLoc`, but (1) It is very difficult due to needing to know the borrow state of a local and (2) There is no difference in gas cost AFAIK. It is done for references out of necessity.
